vowpalwabbit/parse_dispatch_loop.h

```diff
@@ -12,66 +12,63 @@
 
 #include "io/logger.h"
 
-using dispatch_fptr = std::function<void(vw&, const v_array<example*>&)>;
+using dispatch_fptr = std::function<void(vw&, const std::vector<example*>&)>;
 struct io_state;
 
 inline void parse_dispatch(vw& all, dispatch_fptr dispatch)
 {
-  v_array<example*> examples;
+  
   size_t example_number = 0;  // for variable-size batch learning algorithms
 
-  // for substring_to_example
+  // we need to allow reuse of vectors, instead of deleting and recreating them for each example.
+  // used in substring_to_example.
   std::vector<VW::string_view> words_localcpy;
   std::vector<VW::string_view> parse_name_localcpy;
 
   try
   {
     while (!all.example_parser->done)
     {
-
       example* example_ptr = &VW::get_unused_example(&all);
-      examples.push_back(example_ptr);
-
-      if (!all.do_reset_source && example_number != all.pass_length && all.max_examples > example_number &&
-          all.example_parser->reader(&all, examples, words_localcpy, parse_name_localcpy) > 0)
-      {
+      std::vector<example*>* examples = &VW::get_unused_example_vector(&all);
+      (*examples).push_back(example_ptr);
 
-        VW::setup_examples(all, examples);
-        example_number += examples.size();
+      bool is_null_pointer = false;
 
-        dispatch(all, examples);
+      // This logic should be outside the reader, as:
+      // 1. it is common to all types of parsers.
+      // 2. the type of parser might change when the thread is waiting for a pop of io_lines.
+      std::vector<char> *io_lines_next_item = nullptr;
+      {
+        std::lock_guard<std::mutex> lck(all.example_parser->parser_mutex);
+        
+        io_lines_next_item = all.example_parser->io_lines.pop();
 
+        if(io_lines_next_item != nullptr) {
+          all.example_parser->ready_parsed_examples.push(examples);
+        }
       }
-      else
-      { 
 
-        reset_source(all, all.num_bits);
 
-        // to call reset source in io thread
-        all.example_parser->done_with_io.store(true);
-        all.example_parser->can_end_pass.notify_one();
-
-        all.do_reset_source = false;
-        all.passes_complete++;
+      if(io_lines_next_item == nullptr) {
```

Comment:
add a comment mentioning this can only happen after all the parsing is finished (including all the passes)

Reply:
Doesn't this happen at the end of each pass?